proposals/004-health-checks.md

```diff
@@ -0,0 +1,87 @@
+
+We do not want the OpenShift liveness probe to react on errors from the integration because when an integration is in an error state because of one of its connectors fails, this should be reported by other means than Kubernetes status changes. It should be visible on the Syndesis UI instead (i.e. integration health checks).
+
+### Integration health checks (status)
+
+These are Syndesis specific and requires new functionality developed into Apache Camel and as well in Syndesis.
+
+## Implementation
+
+The existing Camel health check from camel-spring-boot (`/health` HTTP REST endpoint) can be used as-is, however as we want Syndesis to be more in control of errors, we would need to ensure that the readiness/liveness probes reports OK even if an integration startup and cannot connect to a remote system, and would be in error.
+
+When Apache Camel startup then Camel routes will startup by default, and as part of their startup then the (route consumers) will often connect to remote systems,
+and in case of connection errors those will be thrown as exception and Camel will fail to startup. Only a handful of Camel components has built-in failover/retry in those situatations.
+
+For Syndesis we would need to be in full control, and therefore we will need to ensure Apache Camel can startup even if there is a connection problem with a remote system. This requires to configure Apache Camel to not auto start the routes, which can be done in Spring Boot by configuring `application.properties` with the following:
+
+    camel.springboot.auto-startup = false
+
+By doing so no routes is automatic startup up, but we would then need to defer starting these routes and do this using another way. Therefore we implement a new `RouteController` which will be responsible for starting up the routes in a controlled manner.
+
+The `RouteController` will be bootstrapped via the `RoutePolicyFactory` where we in the `onInit` callback can register the routes to the `RouteController` which then
+runs as a background thread and orchestrates starting the routes. The `RouteController` will then be responsbible for starting up all these routes, and have support for periodically retry in case a route fails to startup. And only when all routes has been started the `RouteController` has completed starting up all the routes.
+
+We would need to make the discovery and bootstrap of the `RouteController` very easy, which we can do using Camel's existing `RoutePolicyFactory`.
```

Comment:
* [this](http://camel.apache.org/routepolicy.html) is the main doc
* [here](https://github.com/apache/camel/tree/master/camel-core/src/main/java/org/apache/camel/impl/ha) an implementation that delays the startup of the routes as soon as this camel instance becomes the master in a clustered env